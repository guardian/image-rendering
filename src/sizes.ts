// ----- Imports ----- //

import { css, SerializedStyles } from '@emotion/core';
import { Breakpoint, from, breakpoints } from '@guardian/src-foundations/mq';


// ----- Types ----- //

type Size = {
    breakpoint: Breakpoint;
    size: string;
}

type Sizes = {
    mediaQueries: Size[];
    default: string;
}


// ----- Functions ----- //

const sizesAttribute = (sizes: Sizes): string => {
    if (sizes.mediaQueries.length === 0) {
        return sizes.default;
    }

    const queries = sizes.mediaQueries.map(
        query => `(min-width: ${breakpoints[query.breakpoint]}px) ${query.size}`,
    ).join(', ');

    return `${queries}, ${sizes.default}`;
}

const dimensions = (size: string, ratio: number): SerializedStyles => css`
    width: ${size};
    height: calc(${size} * ${ratio});
`;

const styles = (sizes: Sizes, width: number, height: number): SerializedStyles => {
    const ratio = height / width;
    
    return css`
        ${dimensions(sizes.default, ratio)}

        ${sizes.mediaQueries.map(query => css`
            ${from[query.breakpoint]} {
                ${dimensions(query.size, ratio)}
            }
        `)}
    `;
}


// ----- Exports ----- //

export {
    Sizes,
    styles,
    sizesAttribute,
};
