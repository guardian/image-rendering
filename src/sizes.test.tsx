/** @jsx jsx */
// ----- Imports ----- //

import renderer from 'react-test-renderer';
import { breakpoints } from '@guardian/src-foundations/mq';
import { jsx } from '@emotion/core';
import { matchers } from 'jest-emotion';

import { sizesAttribute, Sizes, styles } from './sizes';


// ----- Setup ----- //

expect.extend(matchers);

const noQueries = { mediaQueries: [], default: '30vw' };
const sizes: Sizes = {
    mediaQueries: [
        { breakpoint: 'desktop', size: '20px' },
        { breakpoint: 'tablet', size: '20%' },
    ],
    default: '30vw',
};


// ----- Tests ----- //

describe('sizesAttribute', () => {    
    it('returns the default when there are no media queries', () => {
        expect(sizesAttribute(noQueries)).toBe(noQueries.default);
    });

    it('returns a list of media queries', () => {
        expect(sizesAttribute(sizes)).toMatchSnapshot();
    });
});

describe('styles', () => {
    it('returns the default dimensions when there are no media queries', () => {
        const component = renderer.create(
            <div css={styles(noQueries, 2000, 1500)} />,
        ).toJSON();

        expect(component).toMatchSnapshot();
        expect(component).toHaveStyleRule('width', noQueries.default);
        expect(component).toHaveStyleRule(
            'height',
            `calc(${noQueries.default} * 0.75)`,
        );
    });

    it('returns dimensions for each media query', () => {
        const component = renderer.create(
            <div css={styles(sizes, 600, 300)} />
        ).toJSON();

        expect(component).toMatchSnapshot();
        expect(component).toHaveStyleRule('width', '30vw');
        expect(component).toHaveStyleRule(
            'width',
            '20%',
            { media: `(min-width: ${breakpoints.tablet}px)` },
        );
        expect(component).toHaveStyleRule(
            'height',
            `calc(${sizes.mediaQueries[0].size} * 0.5)`,
            { media: `(min-width: ${breakpoints.desktop}px)` },
        );
    });
});
