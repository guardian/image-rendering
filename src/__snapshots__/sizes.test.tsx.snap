// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sizesAttribute returns a list of media queries 1`] = `"(min-width: 980px) 20px, (min-width: 740px) 20%, 30vw"`;

exports[`styles returns dimensions for each media query 1`] = `
.emotion-0 {
  width: 30vw;
  height: calc(30vw * 0.5);
}

@media (min-width:980px) {
  .emotion-0 {
    width: 20px;
    height: calc(20px * 0.5);
  }
}

@media (min-width:740px) {
  .emotion-0 {
    width: 20%;
    height: calc(20% * 0.5);
  }
}

<div
  className="emotion-0"
/>
`;

exports[`styles returns the default dimensions when there are no media queries 1`] = `
.emotion-0 {
  width: 30vw;
  height: calc(30vw * 0.75);
}

<div
  className="emotion-0"
/>
`;
